src/go/kreq-gen/kafka-request-generator.go

```diff
@@ -0,0 +1,239 @@
+			bytearr = append(bytearr, record.AppendTo(nil)...)
+		}
+	}
+	return bytearr
+}
+
+func randomFill(v reflect.Value, version int16) {
+	switch v.Kind() {
+	case reflect.Ptr:
+		v.Set(reflect.New(v.Type().Elem()))
+		/// This conditional represents setting an nullable field to nil
+		/// happening 50% of the time
+		if makeRandomBool() {
+			randomFill(v.Elem(), version)
+		}
+	case reflect.Struct:
+		for i := 0; i < v.NumField(); i++ {
+			randomFill(v.Field(i), version)
+		}
+		// Special cases, the following byte arrays must be populated with kafka
+		// record & recordBatch formatted data
+		if v.Type().Name() == "ProduceRequestTopicPartition" {
+			recordBatchData := makeRandomKafkaData(version)
+			v.FieldByName("Records").Set(reflect.ValueOf(recordBatchData))
+		} else if v.Type().Name() == "FetchResponseTopicPartition" {
+			recordBatchData := makeRandomKafkaData(version)
+			v.FieldByName("RecordBatches").Set(reflect.ValueOf(recordBatchData))
+		}
+	case reflect.Array:
+		num := v.Len()
+		arr := reflect.New(reflect.ArrayOf(num, v.Type().Elem())).Elem()
+		v.Set(arr)
+		for i := 0; i < num; i++ {
+			randomFill(v.Index(i), version)
+		}
+	case reflect.Slice:
+		num := rand.Intn(10)
+		v.Set(reflect.MakeSlice(v.Type(), num, num))
+		for i := 0; i < num; i++ {
+			randomFill(v.Index(i), version)
+		}
+	case reflect.Bool:
+		v.SetBool(makeRandomBool())
+	case reflect.Uint8:
+		v.SetUint(uint64(makeRandomInt64()))
+	case reflect.Int8:
+		fallthrough
+	case reflect.Int16:
+		fallthrough
+	case reflect.Int32:
+		fallthrough
+	case reflect.Int64:
+		v.SetInt(makeRandomInt64())
+	case reflect.String:
+		v.SetString(makeRandomString())
+	case reflect.Map:
+		/// Only case where Map is an acceptable type
+		tags := makeRandomTags()
+		/// The tags structure is an unexported private field
+		/// Use `unsafe.Pointer` to be able to set it
+		rf := reflect.NewAt(v.Type(), unsafe.Pointer(v.UnsafeAddr())).Elem()
+		rf.Set(reflect.MakeMapWithSize(reflect.TypeOf(map[uint32][]byte{}), tags.Len()))
+		tags.Each(func(key uint32, element []byte) {
+			rf.SetMapIndex(reflect.ValueOf(key), reflect.ValueOf(element))
+		})
+	default:
+		/// TODO: Special case produce_request / records, must be formatted record batches
+		fmt.Printf("Invalid type: %s\n", v.Kind())
+		os.Exit(1)
+	}
+}
+
+func randomFillType(p interface{}, version int16) {
+	randomFill(reflect.ValueOf(p).Elem(), version)
+}
+
+func main() {
+	rand.Seed(time.Now().UnixNano())
+	api := flag.Int("api", 0, "API key of request to generate")
+	version := flag.Int("version", 0, "API version of generated request")
+	isRequest := flag.Bool("is-request", true, "Generate kafka request if true, response if false")
+	outputAsBinary := flag.Bool("binary-output", true, "Print result as binary")
+	var result []byte
+	flag.Parse()
+	if *isRequest {
+		req := kmsg.RequestForKey(int16(*api))
+		randomFillType(req, int16(*version))
+		reflect.ValueOf(req).Elem().FieldByName("Version").SetInt(int64(*version))
```

Comment:
The req has a function, `SetVersion`, that can replace this and the duplication below for resp